### DISCUSSION TITLE
[Kubernetes][Kubebuilder] New Golang plugin to help Operator authors skill up

### DISCUSSION THREAD

**Discussion by nate-double-u** [cncf/mentoring#571] [Google Summer of Code 2022]
Your goal is to develop a [Kubebuilder Plugin](https://book.kubebuilder.io/plugins/plugins.html) which generates the files with all possible common desired source code implementation for an Operator Author to achieve the goal of deploying an Operand(image/Pod) following the [Operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) and common best practices and recommendations such as using [StatusConditionals](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties), tests, etc. Note that we can begin implementing this plugin with basic implementation and then start to provide many follow-ups as much we wish to grow it incrementally. This plugin can help a lot of Operator Authors save time and give good direction and start to point to them. You can begin by following, for example, the quick tutorial [Golang Operator](https://sdk.operatorframework.io/docs/building-operators/golang/tutorial/) and the document [Common recommendations and suggestions](https://sdk.operatorframework.io/docs/best-practices/common-recommendation/) to have an idea of how Operators works and what code this plugin would generate by default. It is ideal for those who are looking to know more about [Operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) and its good practices, how to develop tests and ensure the quality and maintainability of solutions.

Mentor(s): Camila Macedo (@camilamacedo86) and Rashmi Gottipati(@rashmigottipati)

https://github.com/cncf/mentoring/blob/main/summerofcode/2022.md#new-golang-plugin-to-help-operator-authors-skill-up-size-medium

**Comment by camilamacedo86** [comment]
Following are a summary of the questions/answers made so far and that you might have. I hope that helps you out. 

# FAQ:

## 💁‍♀️  What does Kubebuilder? How it is used?

Kubebuilder is often used as a CLI tool that generates projects and provides features to help people create "Operator Solutions".
"Operator solutions" are projects which follow the [Operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) and [extend the K8S API](https://kubernetes.io/docs/tasks/extend-kubernetes/) and has the purpose to manage solutions on K8s clusters. 

**TL'DR:** 
> All Operators are made of [controllers](https://github.com/kubernetes-sigs/controller-runtime), but not all sets of controllers are operators. Operators are sets of controllers that encode human operational knowledge for a specific application (e.g. the Prometheus Operator).  Controllers are more general -- for instance, the namespace service account controller isn't an operator because it doesn't have knowledge of a specific application -- it's very general. Kubebuilder helps constructs sets of controllers.  Some of those are Operators, some might not be.

Also, Kubebuilder can be used as a lib to help other projects extend their features and/or create nice [plugins](https://book.kubebuilder.io/plugins/plugins.html) which can be used as specific helpers. For example, [Operator-SDK](https://github.com/operator-framework/operator-sdk) uses Kubebuilder and scaffolds the same plugins/projects but also provide additional features and others plugins on top. 

See: 
- [Readme Overview](https://github.com/kubernetes-sigs/kubebuilder#kubebuilder)
- [Readme Motivatios](https://github.com/kubernetes-sigs/kubebuilder#motivation)
- [Who is this for](https://book.kubebuilder.io/introduction.html#who-is-this-for)

## 💡  Examples of its usage:

See https://operatorhub.io/ to have a better idea

## ❓ Why Operators?

By using operators, it’s possible not only to provide all expected resources but also to manage them dynamically, programmatically, and at execution time. To illustrate this idea, imagine if someone accidentally changed a configuration or removed a resource by mistake; in this case, the operator could fix it without any human intervention.

## What is the difference between Kubebuilder and SDK? 

See: https://sdk.operatorframework.io/docs/faqs/#what-are-the-the-differences-between-kubebuilder-and-operator-sdk

## 💁  Why should we do the proposed new plugin? What are its motivations? 

Many authors create a project to deploy/manage an image ( app ) on the cluster then, with this new proposed plugin that you will help us to have we ought to be able to scaffold all that is required to get the deployment of this image managed on the cluster.

So, that means ( e.g. ):

- scaffold in the reconcile the code to create the Deployment resource of the image
- scaffold by default the code that is required to properly use the [StatusConditionals](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#typical-status-properties)
- scaffold by default the tests required to ensure that this implementation works

So that, we can help users to skill up in the good practices and how to do the things. Also, we can help users deliver their needs faster and by spending lower effort. 

## 💁‍♀️  How to know more about the plugins and how to write them? 

See the doc: https://book.kubebuilder.io/plugins/plugins.html

## 🚀 How can I begin to be familiarized with the Project? 

OpenSource projects usually have a Contribution guide, see [CONTRIBUTING](https://github.com/kubernetes-sigs/kubebuilder/blob/master/CONTRIBUTING.md) and also [VERSIONING](https://github.com/kubernetes-sigs/kubebuilder/blob/master/VERSIONING.md) for further details.

Also, feel free to check the open issues in the project and began to work and contribute with anything that you wish. 
Your help and contributions 🥇 are very welcome. 

The maintainers actively manage the issues list and try to highlight issues suitable for newcomers (You can filter by [good+first+issue](https://github.com/kubernetes-sigs/kubebuilder/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) also contributions to help on the docs can fit well for the first ones) but you are free to help as please you. The project follows the typical GitHub pull request model. Before starting any work, please either comment on an existing issue or file a new one.

## ✋ How to get help? 

Feel free to ping us and join the #kubebuilder channel in the kubernetes.slack.com slack. 
You also can open issues in its repo if you see that fits. 

## How can I provide the flags for the plugins? Where they are defined?

In the plugin layout, we have some interfaces for the commands which like `kubebulder init [OPTIONS]`, `kubebulder create API [OPTIONS]`, `kubebulder edit [OPTIONS]`, `kb create webhooks [OPTIONS]`.

Then, for each plugin, we implement these interfaces (subCommands), e.g.:
- for the declarative plugin which can be called only when we create the project and when we create an api we just implement both subCommands, see: https://github.com/kubernetes-sigs/kubebuilder/tree/master/pkg/plugins/golang/declarative/v1
- However, for the go/v3 plugin, we implement all, see: https://github.com/kubernetes-sigs/kubebuilder/tree/master/pkg/plugins/golang/v3

Then see that we are biding the flags for each subCommand implementation, e.g.
- flags when for when we run the command `kb init` using the go/v3 plugin: [here](https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/plugins/golang/v3/init.go#L79-L94)
- flags when for when we run the command `kb create api` [here](https://github.com/kubernetes-sigs/kubebuilder/blob/master/pkg/plugins/golang/v3/api.go#L96-L120)

## 💡  How to skill up and getting started with?

- You can follow the Kubebuilder QuickStart and tutorial: https://book.kubebuilder.io/quick-start.html
- (faster) You can check SDK docs as well, see that it also have tutorials. Then, note that the [Memcached sample](https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v3/memcached-operator) can be followed and built with Kubebuilder CLI tool, see: https://sdk.operatorframework.io/docs/building-operators/golang/tutorial/
- Do not forget to check out the whole Plugin section: https://book.kubebuilder.io/plugins/plugins.html

## Who are the students that reached out so far: 

| slack | GitHub | URL | 
| --- | --- | --- | 
| Ashwin P | @Ashwin901 | https://github.com/Ashwin901 | 
| Nivedita Prasad |  @Nivedita-coder | https://github.com/Nivedita-coder | 
| Nikhil Sharma | @NikhilSharmaWe | https://github.com/NikhilSharmaWe | 
| Ankit Kurmi | @Ankit152 | https://github.com/Ankit152 |
| rahul sawra | @rahulii | https://github.com/rahulii | 
| Sangwoo Bae | @sangwoobae | https://github.com/sangwoobae | 
| TianYi(Tony) | ? | ? | ? |
| parul_sahoo | @parul5sahoo | https://github.com/parul5sahoo | 
| Sayantan Mondal  | @Saty248 |  https://github.com/Saty248 |
